### ISSUE TITLE
[Technical Question] Why are there 6001 outputs for yolov5 despite having number of boxes as an output?

### ISSUE THREAD

**Issue by timothylimyl** [wang-xinyu/tensorrtx#600]
There's a total of 6001 outputs but the amount of "number of boxes" are used for post-processing:

```
# Get the num of boxes detected
num = int(output[0])
# Reshape to a two dimentional ndarray
pred = np.reshape(output[1:], (-1, 6))[:num, :]

```

If it outputs the number of boxes it cares about, what is the purpose of the rest of the outputs? For example, if `num==10` then `6000 - (10*6)` outputs are useless in the post-processing. It just seems weird to me that the detection model can output the number of boxes it cares about but also outputs all the information it does not need. 

Anyways, I test it without the `[:num,:]` filter and it works the same for me but obviously, the processing is slightly more. Mainly, I do not understand how does the model know the `number of boxes` and what is used to determine this?

**TL:DR: Confused about Yolov5 raw model outputs, anyone can help clarify?**

**Comment by wang-xinyu** [comment]
If walking through yololayer.cu, you can get basic understanding.
Plugin need a fixed output dimension in building phase. 6000 is because MAX_OUTPUT_NUM is 1000, and

**Comment by timothylimyl** [comment]
@wang-xinyu  thank you for your reply, appreciate it. I am now clear on the fixed output dimension needed as you need to pre-allocate space on the GPU.

However, the number of boxes wouldn't be flexible right? The yolov5 model will output a **limited** number of boxes based on the number of anchor boxes assigned per grid cell?  My initial thought was that the number of boxes predicted will be set during training and inference code will be forced to follow.

**Comment by wang-xinyu** [comment]
Right, the max amount of objects that can be detected is determined by the output dims of the last three conv layers.

**Comment by timothylimyl** [comment]
ok, thank you, have a good one.
